### ISSUE TITLE
Weird behavior with branches

### ISSUE THREAD

**Issue by jcbertin** [ingydotnet/git-subrepo#387]
I have weird behavior when pushing locally created branches to remote with `git subrepo push --branch`: sometimes the commits log is just fine, but sometimes it fails using exactly the same commands.
I made a snippet to expose that: 
[TestSubrepo.txt](https://github.com/ingydotnet/git-subrepo/files/2643120/TestSubrepo.txt)

The commit log containing the subrepo looks like that: 
```
~/Test$ git log --graph --branches
* commit c43f71bc1c9d3938cd7cc6f21f75125934dba5a0 (HEAD -> master, origin/master)
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:13 2018 +0100
|
|     git subrepo push TestSubrepo
|
|     subrepo:
|       subdir:   "TestSubrepo"
|       merged:   "bb9a668"
|     upstream:
|       origin:   "../remotes/TestSubrepo"
|       branch:   "master"
|       commit:   "bb9a668"
|     git-subrepo:
|       version:  "0.4.0"
|       origin:   "https://github.com/ingydotnet/git-subrepo"
|       commit:   "5d6aba9"
|
* commit 45e0f20cc05cfa8ec38f0e5ff3caaa12da9b7710
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     4th commit.
|
| * commit f1821533c0ce918f398a0d2cebbf2a02ca45a1d3 (origin/toto, toto)
| | Author: Jean-Charles Bertin
| | Date:   Tue Dec 4 02:58:13 2018 +0100
| |
| |     git subrepo push --branch=toto TestSubrepo
| |
| |     subrepo:
| |       subdir:   "TestSubrepo"
| |       merged:   "65f9d45"
| |     upstream:
| |       origin:   "../remotes/TestSubrepo"
| |       branch:   "toto"
| |       commit:   "65f9d45"
| |     git-subrepo:
| |       version:  "0.4.0"
| |       origin:   "https://github.com/ingydotnet/git-subrepo"
| |       commit:   "5d6aba9"
| |
| * commit bb5a0c5b3c0a3b39ce020ac0c87430bfaeffbed7
| | Author: Jean-Charles Bertin
| | Date:   Tue Dec 4 02:58:12 2018 +0100
| |
| |     3rd commit.
| |
| * commit eef5e24861e7456885a50be4842836eabe666eb5
|/  Author: Jean-Charles Bertin
|   Date:   Tue Dec 4 02:58:12 2018 +0100
|
|       2nd commit.
|
* commit 03e9b0586faaa5f338cb4bf5aa29bcf715387924
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     First commit.
|
* commit f9fef50f773756f55b374b66bc84bb992ab5d806
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     git subrepo clone ../remotes/TestSubrepo
|
|     subrepo:
|       subdir:   "TestSubrepo"
|       merged:   "2a66312"
|     upstream:
|       origin:   "../remotes/TestSubrepo"
|       branch:   "master"
|       commit:   "2a66312"
|     git-subrepo:
|       version:  "0.4.0"
|       origin:   "https://github.com/ingydotnet/git-subrepo"
|       commit:   "5d6aba9"
|
* commit a6200d28e5bc5a19a027b3180579dfadaf80e26e
  Author: Jean-Charles Bertin
  Date:   Tue Dec 4 02:58:11 2018 +0100

      Initial commit.
```

When running these commands multiple times, the commits log in the remote is not always the same ! Sometimes, it is right:
```
~/TestSubrepo$ git log --graph --remotes
* commit dd922c046baee9893425c792662c291ef454c1bf (origin/toto)
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:16 2018 +0100
|
|     3rd commit.
|
* commit 3222196be4e5595e0aa905e113584f162ac12520
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:16 2018 +0100
|
|     2nd commit.
|
| * commit dc5332a5c7c809b1992b691f017e4678348aa325 (origin/master)
|/  Author: Jean-Charles Bertin
|   Date:   Tue Dec 4 02:58:16 2018 +0100
|
|       4th commit.
|
* commit 1baf2a49b8091a5f013a24dc3a05acc3b052c746
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:16 2018 +0100
|
|     First commit.
|
* commit e679d4a4c70df197d1bb241244b0fbefb0940f90 (HEAD -> master)
  Author: Jean-Charles Bertin
  Date:   Tue Dec 4 02:58:16 2018 +0100

      Initial commit.
```

But sometimes, it goes wrong:
```
~/TestSubrepo$ git log --graph --remotes
* commit 65f9d45d71a2899798cbecb6f8d8166caa67f31b (origin/toto)
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     3rd commit.
|
* commit e692406f881bc15d3eef3c17e84dd8306fef8c42
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     2nd commit.
|
* commit 23c097a8b6905da52d1521044fc45424528f57a0
| Author: Jean-Charles Bertin
| Date:   Tue Dec 4 02:58:12 2018 +0100
|
|     First commit.
|
| * commit bb9a6682c21a0e6f9c3d4c97099372646b0697fd (origin/master)
| | Author: Jean-Charles Bertin
| | Date:   Tue Dec 4 02:58:12 2018 +0100
| |
| |     4th commit.
| |
| * commit a41d3e5888b2b449d1f8919c7e842a5326133719
|/  Author: Jean-Charles Bertin
|   Date:   Tue Dec 4 02:58:12 2018 +0100
|
|       First commit.
|
* commit 2a6631237da0fda559e729dba0f5370be5820829 (HEAD -> master)
  Author: Jean-Charles Bertin
  Date:   Tue Dec 4 02:58:11 2018 +0100

      Initial commit.
```

Thus, using git-subrepo 0.4.0 and git 2.19.2.
Any help is welcome. Did I make something wrong?

**Comment by grimmySwe** [comment]
@jcbertin I have tested your script and reproduced the issue locally, I haven't been able to figure out why it happens though. I might be tempted to say that the "wrong" case is actually what I was expected, but it's really strange that it behaves differently!
Will look into this further and see if I can find anything.

**Comment by grimmySwe** [comment]
@jcbertin Now I have found the "cause" of the difference. git-subrepo recreates commits and depending on timing these commits will be equivalent or not. Hash is based on time stamps for both author and commit and if they happen the same second, it will be find the same commit, if not it will be separate ones.

I will look into if I can set the commit date to the same to create the same behavior every time.

**Comment by grimmySwe** [comment]
If you want the same behavior in your script, add sleep 1 between pushing the subrepos. That way you get the "expected" outcome.
As subrepo creates new commits it will generate different commits unless we set it hard and we do want the commit data to be set to when the committer actually used subrepo, not when the original commit was authored. It should work the same way as cherry-picking.

**Comment by admorgan** [comment]
I believe this was fixed with c9895df36d571f21934eea63a9210438d32fce9b. If you can reproduce this please reopen this.

**Comment by jcbertin** [comment]
@admorgan: actually the behavior is now coherent but it produces always the wrong tree!
Did I miss something?

```
* commit b8c0a3b4631f93ab0fcb86f237c7267ac512fee2 (origin/toto)
| Author: Jean-Charles Bertin
| Date:   Tue Jan 21 07:48:55 2020 +0100
|
|     3rd commit.
|
* commit 47a2dde45ebf22137e023e5bb82edcba5d757a1b
| Author: Jean-Charles Bertin
| Date:   Tue Jan 21 07:48:54 2020 +0100
|
|     2nd commit.
|
* commit 68ca5136fbf9e84417d560d86b791b9e7effc80c
| Author: Jean-Charles Bertin
| Date:   Tue Jan 21 07:48:53 2020 +0100
|
|     First commit.
|
| * commit 54a8be9314e37f6d249d34734742b7bc0576d67a (origin/master)
| | Author: Jean-Charles Bertin
| | Date:   Tue Jan 21 07:48:56 2020 +0100
| |
| |     4th commit.
| |
| * commit ffd8fb9a0ee9a7221b053828fafdcc474a4d2fdd
|/  Author: Jean-Charles Bertin
|   Date:   Tue Jan 21 07:48:53 2020 +0100
|
|       First commit.
|
* commit 563b81473ab89d422c08c8c6bfa8b612b0a673d9 (HEAD -> master)
  Author: Jean-Charles Bertin
  Date:   Tue Jan 21 07:48:52 2020 +0100

      Initial commit.
```
